File: teamapps-ux/src/main/java/org/teamapps/ux/component/collapsible/Collapsible.java
/*-
 * ========================LICENSE_START=================================
 * TeamApps
 * ---
 * Copyright (C) 2014 - 2024 TeamApps.org
 * ---
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * 
 *      http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * =========================LICENSE_END==================================
 */
package org.teamapps.ux.component.collapsible;

import org.teamapps.dto.UiCollapsible;
import org.teamapps.dto.UiEvent;
import org.teamapps.event.Event;
import org.teamapps.icons.Icon;
import org.teamapps.ux.component.AbstractComponent;
import org.teamapps.ux.component.Component;

public class Collapsible extends AbstractComponent {

	public final Event<Boolean> onCollapseStateChanged = new Event<>();

	private Icon<?, ?> icon;
	private String caption;
	private Component content;
	private boolean collapsed;

	public Collapsible() {
		this(null, null, null);
	}

	public Collapsible(Icon<?, ?> icon, String caption) {
		this(icon, caption, null);
	}

	public Collapsible(Icon<?, ?> icon, String caption, Component content) {
		this.icon = icon;
		this.caption = caption;
		this.content = content;
	}

	@Override
	public UiCollapsible createUiComponent() {
		UiCollapsible ui = new UiCollapsible();
		mapAbstractUiComponentProperties(ui);
		ui.setIcon(getSessionContext().resolveIcon(icon));
		ui.setCaption(caption);
		ui.setContent(content != null ? content.createUiReference() : null);
		ui.setCollapsed(collapsed);
		return ui;
	}

	@Override
	public void handleUiEvent(UiEvent event) {
		switch (event.getUiEventType()) {
			case UI_COLLAPSIBLE_COLLAPSE_STATE_CHANGED: {
				UiCollapsible.CollapseStateChangedEvent ce = (UiCollapsible.CollapseStateChangedEvent) event;
				onCollapseStateChanged.fire(ce.getCollapsed());
			}
		}
	}

	public Component getContent() {
		return content;
	}

	public void setContent(Component content) {
		this.content = content;
		queueCommandIfRendered(() -> new UiCollapsible.SetContentCommand(getId(), content != null ? content.createUiReference() : null));
	}

	public Icon<?, ?> getIcon() {
		return icon;
	}

	public void setIcon(Icon<?, ?> icon) {
		this.icon = icon;
		queueCommandIfRendered(() -> new UiCollapsible.SetIconAndCaptionCommand(getId(), getSessionContext().resolveIcon(icon), caption));
	}

	public String getCaption() {
		return caption;
	}

	public void setCaption(String caption) {
		this.caption = caption;
		queueCommandIfRendered(() -> new UiCollapsible.SetIconAndCaptionCommand(getId(), getSessionContext().resolveIcon(icon), caption));
	}

	public boolean isCollapsed() {
		return collapsed;
	}

	public void setCollapsed(boolean collapsed) {
		this.collapsed = collapsed;
		queueCommandIfRendered(() -> new UiCollapsible.SetCollapsedCommand(getId(), collapsed));
	}
}

File: client/projector-collapsible/src/main/ts/Collapsible.ts
/*-
 * ========================LICENSE_START=================================
 * TeamApps
 * ---
 * Copyright (C) 2014 - 2025 TeamApps.org
 * ---
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * =========================LICENSE_END==================================
 */

import {
    AbstractComponent,
    Component,
    parseHtml, ProjectorEvent,
    ServerObjectChannel,
    toggleElementCollapsed
} from "projector-client-object-api";
import {
    DtoCollapsible,
    DtoCollapsible_CollapseStateChangedEvent,
    DtoCollapsibleCommandHandler,
    DtoCollapsibleEventSource
} from "./generated";

export class Collapsible extends AbstractComponent<DtoCollapsible> implements DtoCollapsibleCommandHandler, DtoCollapsibleEventSource {

    readonly onCollapseStateChanged: ProjectorEvent<DtoCollapsible_CollapseStateChangedEvent> = new ProjectorEvent();

    private $main: HTMLDivElement;
    private $icon: HTMLElement;
    private $caption: HTMLElement;
    private $body: HTMLElement;
    private $expander: Element;

    constructor(config: DtoCollapsible, serverObjectChannel: ServerObjectChannel) {
        super(config);
        this.$main = parseHtml(`<div class="Collapsible">
	<div class="collapsible-header">
		<div class="expander teamapps-expander ${config.collapsed ? '' : 'expanded'}"></div>
		<div class="icon img img-16"></div>
		<div class="caption"></div>
	</div>
	<div class="collapsible-body"></div>
</div>`);
        this.$expander = this.$main.querySelector(":scope .expander");
        this.$icon = this.$main.querySelector(":scope .icon");
        this.$caption = this.$main.querySelector(":scope .caption");
        this.$body = this.$main.querySelector(":scope .collapsible-body");

        this.setIcon(config.icon);
        this.setCaption(config.caption);
        this.setContent(config.content);
        this.setCollapsed(config.collapsed);

        this.$expander.addEventListener("click", evt => {
            let collapsed = !this.config.collapsed;
            this.setCollapsed(collapsed);
            this.onCollapseStateChanged.fire({collapsed});
        });
    }

    public doGetMainElement(): HTMLElement {
        return this.$main;
    }

    setContent(content: unknown): void {
        this.$body.innerHTML = "";
        if (content != null) {
            this.$body.append((content as Component).getMainElement());
        }
    }

    setCollapsed(collapsed: boolean): any {
        this.config.collapsed = collapsed;
        this.$main.classList.toggle("collapsed", collapsed);
        this.$expander.classList.toggle("expanded", !collapsed);

        this.$main.classList.add("expanding")
        toggleElementCollapsed(this.$body, collapsed, 300, "hidden", () => {
            this.$main.classList.remove("expanding");
        });
    }

    setCaption(caption: string): any {
        this.config.caption = caption;
        this.$caption.innerText = caption;
    }

    setIcon(icon: string): any {
        this.config.icon = icon;
        this.$icon.style.backgroundImage = `url('${icon}')`;
    }

}
